pkg/sql/schemachanger/scbuild/internal/scbuildstmt/drop_index.go

```diff
@@ -0,0 +1,438 @@
+	explicitColNames := make([]string, len(explicitColIDs))
+	for i, colID := range explicitColIDs {
+		colElms := b.ResolveColumnByID(ie.TableID, colID, ResolveParams{
+			IsExistenceOptional: false,
+			RequiredPrivilege:   privilege.CREATE,
+		})
+		_, _, cne := scpb.FindColumnName(colElms)
+		if cne == nil {
+			panic(fmt.Sprintf("No column name is found for column ID %v", colID))
+		}
+		explicitColNames[i] = cne.Name
+	}
+
+	colIDs := make(descpb.ColumnIDs, 0, len(explicitColIDs))
+	for i := range explicitColNames {
+		if ie.Sharding == nil || !ie.Sharding.IsSharded || explicitColNames[i] != ie.Sharding.Name {
+			colIDs = append(colIDs, explicitColIDs[i])
+		}
+	}
+	return colIDs
+}
+
+// isValidReferencedUniqueConstraint return true if `e` is a unique index and this index's
+// explicit columns are a permutation of `referencedColIDs`.
+func isValidReferencedUniqueConstraint(
+	b BuildCtx, ie *scpb.Index, referencedColIDs []tree.ColumnID,
+) bool {
+	return ie.IsUnique && !ie.IsPartial &&
+		explicitColumnIDsWithoutShardColumn(b, ie).PermutationOf(referencedColIDs)
+}
+
+// uniqueConstraintHasReplacementCandidate returns true if `elms` contains an index
+// that can serve as a replacement candidate for the to-be-dropped secondary index `sie`, which
+// references columns `referencedColumnIDs`.
+func uniqueConstraintHasReplacementCandidate(
+	b BuildCtx,
+	elms ElementResultSet,
+	sie *scpb.SecondaryIndex,
+	referencedColumnIDs []descpb.ColumnID,
+) bool {
+	result := false
+
+	// Check all indexes (both primary and secondary) to see if we can find a replacement candidate.
+	elms.ForEachElementStatus(func(current scpb.Status, target scpb.TargetStatus, e scpb.Element) {
+		if result {
+			return
+		}
+
+		switch t := e.(type) {
+		case *scpb.SecondaryIndex:
+			if t.IndexID != sie.IndexID && isValidReferencedUniqueConstraint(b, &t.Index, referencedColumnIDs) {
+				result = true
+			}
+		case *scpb.PrimaryIndex:
+			if isValidReferencedUniqueConstraint(b, &t.Index, referencedColumnIDs) {
+				result = true
+			}
+		}
+	})
+
+	return result
+}
+
+// Attempt to drop all views that depend on the to be dropped index if CASCADE.
+// Panic if there is a dependent view but drop behavior is not CASCADE.
+func mayDropDependentViews(
+	b BuildCtx, toBeDroppedSecondaryIndexElms ElementResultSet, dropBehavior tree.DropBehavior,
+) {
+	_, _, sie := scpb.FindSecondaryIndex(toBeDroppedSecondaryIndexElms)
+	_, _, sine := scpb.FindIndexName(toBeDroppedSecondaryIndexElms)
+	if sie == nil || sine == nil {
+		panic("invalid toBeDroppedSecondaryIndexElms; does not contain a secondary index element.")
+	}
+
+	scpb.ForEachView(b.BackReferences(sie.TableID), func(current scpb.Status, target scpb.TargetStatus, ve *scpb.View) {
+		for _, forwardRef := range ve.ForwardReferences {
+			if forwardRef.IndexID == sie.IndexID {
+				// This view depends on the to-be-dropped index;
+				if dropBehavior != tree.DropCascade {
+					// Get view name for panic message
```

Comment:
I think it's cleaner to just say `for the error message`. The fact that the error is panicked internally in this layer is an implementation detail.